clang/lib/Sema/SemaSYCLDeclAttr.cpp

```diff
@@ -2559,43 +2613,98 @@ void SemaSYCL::handleSYCLAddIRAttributesFunctionAttr(Decl *D,
+  if (NewAttr.getAttributeFilter() != ExistingAttr.getAttributeFilter() ||
+      hasConflictingSYCLAddIRAttributes(
+          NewAttr.getAttributeNameValuePairs(Context),
+          ExistingAttr.getAttributeNameValuePairs(Context))) {
     S.Diag(ExistingAttr.getLoc(), diag::err_duplicate_attribute) << &NewAttr;
     S.Diag(NewAttr.getLoc(), diag::note_conflicting_attribute);
     return true;
   }
   return false;
 }
 
+template <typename AddIRAttrT>
+static AddIRAttrT *getMergedSYCLAddIRAttribute(const AddIRAttrT &Attr1,
+                                               const AddIRAttrT &Attr2,
+                                               SemaSYCL &S) {
+  ASTContext &Context = S.getASTContext();
+  bool AttrHasFilterList = Attr1.hasFilterList();
+
+  // Get the vectors of name-value-pairs here so we can create string references
+  // to them for the map.
+  llvm::SmallVector<std::pair<std::string, std::string>, 4> Attr1NVPairs =
+      Attr1.getAttributeNameValuePairs(Context);
+  llvm::SmallVector<std::pair<std::string, std::string>, 4> Attr2NVPairs =
+      Attr2.getAttributeNameValuePairs(Context);
+
+  // Collect all the unique attribute names and their corresponding values. This
+  // relies on the uniqueness having been confirmed first and that the
+  // attributes appear in the same order as in the name-value-pairs.
+  llvm::SmallMapVector<StringRef, std::pair<Expr *, Expr *>, 4> AttrExprByName;
+  for (const auto &[Attr, NVPairs] : {std::make_pair(Attr1, Attr1NVPairs),
+                                      std::make_pair(Attr2, Attr2NVPairs)}) {
+    for (size_t I = 0; I < NVPairs.size(); ++I) {
+      AttrExprByName[NVPairs[I].first] = std::make_pair(
+          (Attr.args_begin() + AttrHasFilterList)[I],
+          (Attr.args_begin() + AttrHasFilterList + (Attr.args_size() / 2))[I]);
+    }
+  }
+
+  // Create a list of new arguments, starting with the filter if present.
+  llvm::SmallVector<Expr *, 4> NewArgs;
+  NewArgs.resize(AttrExprByName.size() * 2 + AttrHasFilterList);
+  if (AttrHasFilterList)
+    NewArgs[0] = *Attr1.args_begin();
+
+  // Then insert all the unique attributes found previously.
+  for (size_t I = 0; I < AttrExprByName.size(); ++I) {
+    const std::pair<Expr *, Expr *> &Exprs = AttrExprByName.begin()[I].second;
+    NewArgs[I + AttrHasFilterList] = Exprs.first;
+    NewArgs[I + AttrExprByName.size() + AttrHasFilterList] = Exprs.second;
+  }
+
+  return AddIRAttrT::Create(Context, NewArgs.data(), NewArgs.size(), Attr1);
+}
+
 SYCLAddIRAttributesFunctionAttr *SemaSYCL::mergeSYCLAddIRAttributesFunctionAttr(
     Decl *D, const SYCLAddIRAttributesFunctionAttr &A) {
   if (const auto *ExistingAttr =
           D->getAttr<SYCLAddIRAttributesFunctionAttr>()) {
-    checkSYCLAddIRAttributesMergeability(A, *ExistingAttr, *this);
-    return nullptr;
+    if (checkSYCLAddIRAttributesMergeability(A, *ExistingAttr, *this))
+      return nullptr;
+
+    D->dropAttr<SYCLAddIRAttributesFunctionAttr>();
```

Comment:
I don't fully recall how this code works but are you required to drop here if you're dropping in the add function?